k2/addons/k2-serial/zserial/serial.py

```diff
@@ -54,6 +55,75 @@ def serial_exec(sut, messagebus):
     return SerialClient(messagebus, sut.entity, sut.serial.timeout, sut.serial.prompt)
 
 
+@requires(sut='Sut', can=['serial'])
+@requires(messagebus='MessageBus')
+@requires(serial_connection='SerialConnection', instance=True)
+@component(name='RawSerialPort', can=['serial'], provided_by_extension='zserial')
+class RawSerialPort(object):
+    """
+    Access the serial port in raw byte mode.
+
+    Temporarily suspend the packetized log line handling to provide raw byte
+    access to the serial port.
+
+    Example:
+
+    .. code-block:: python
+
+        @requires(serial='RawSerialPort')
+        @requires(exec='Exec', can=['serial'])
+        def test_xmodem_recv(serial,exec):
+            with open('/dev/null', 'wb') as f:
+                exec.send_line('sz -X --1k -q /proc/cpuinfo',
+                    prefix_output=False,
+                    endmark='Give your local XMODEM receive command now.')
+                with serial.suspended() as ser:
+                    def getc(size, timeout=1):
+                        return ser.read(size) or None
+
+                    def putc(data, timeout=1):
+                        return ser.write(data)
+
+                    modem = XMODEM1k(getc, putc)
+                    modem.recv(f)
+
+        @requires(serial='RawSerialPort')
+        @requires(exec='Exec', can=['serial'])
+        def test_xmodem_send(serial,exec):
+            with open('/proc/version', 'rb') as f:
+                exec.send_line('rz -X -q -y /dev/null',
+                    prefix_output=False,
+                    endmark='rz: ready to receive /tmp/foo')
+                with serial.suspended() as ser:
+                    def getc(size, timeout=1):
+                        return ser.read(size) or None
+
+                    def putc(data, timeout=1):
+                        return ser.write(data)
+
+                    modem = XMODEM1k(getc, putc)
+                    modem.send(f)
```

Comment:
This actually works =)

If the K2 interface addon for file transfer happen to re-appear, then I imagine XMODEM being just one of the ways to transfer files to a SUT.

Reply:
I agree that we need a file-transfer mechanism that is done using a component-interface.
I will get back to you on this in a couple of days. I might be able to dig up the HTTP-server based image transfer from upstream.

Reply:
That would be awesome! :smiley:

Reply:
@per-bohlin, any update on the HTTP-server based image transfer from upstream?

Reply:
Sorry, I've been very busy the last couple of weeks.
I'll prioritize to get the reviews done, then I'll look into finding the HTTP-server.